### PR TITLE
Fence historical dispatch during cold Matrix sync

### PR DESCRIPTION
**Pull request by basnijholt**
## Summary

- Fence historical text, edit, media, reaction, and approval callbacks whenever Matrix sync starts without reusable continuity.
- Keep events at or after one fixed runtime-start cutoff eligible while preserving cache certification, state callbacks, and explicit interrupted-turn recovery.
- Apply the same fixed cutoff to decrypt-failure notices so replayed encrypted history stays silent without hiding new failures.

## Tests

- Added callback-level regression coverage for classic and Sliding Sync cold windows, restored checkpoints, retries, token resets, missing timestamps, missing continuation tokens, and decrypt-notice floors.
- Ran 146 focused sync, cache, decrypt, retry, auto-resume, startup, and import-boundary tests.
- Ran the full suite: 12,188 passed and 54 skipped.
- Ran all pre-commit hooks.

## Summary by Sourcery

Fence Matrix cold-sync historical events from user-visible dispatch while preserving cache certification and decrypt-failure notice behavior around a fixed startup cutoff.

New Features:
- Add a startup-based historical-dispatch fence that blocks cold-sync replayed events from entering normal message, reaction, media, and approval handling during tokenless windows.

Enhancements:
- Track a fixed startup cutoff timestamp and reuse it across continuity retries and sync-token resets to distinguish historical from live events.
- Tie historical-dispatch fencing into sync certification, sliding-sync startup, and unknown-position errors so fencing automatically arms or disarms with continuity state.
- Extend decrypt-failure notice flooring to accept an explicit timestamp and enforce a monotonic per-user/room floor to keep replayed encrypted history silent without hiding new failures.

Tests:
- Add comprehensive regression tests around tokenless and sliding Matrix sync windows, restored checkpoints, missing continuation tokens, certification resets, timestamp edge cases, and reaction/approval side effects.
- Add tests ensuring decrypt-failure notice floors interact correctly with the startup cutoff so post-start failures still surface while older failures remain suppressed.

### CODE REVIEW

File: src/mindroom/bot.py [hunk 1]
```diff
@@ -343,6 +346,8 @@ def __init__(
         self._last_sync_monotonic = None
         self._first_sync_done = False
         self._sync_shutting_down = False
+        self._startup_cutoff_ms = int(time.time() * 1000)
+        self._historical_dispatch_fence_armed = False
```
[the remark this answers is not in the document]
Comment (basnijholt) [reply src/mindroom/bot.py]: Not changing this. The approved invariant explicitly captures the cutoff when the bot/runtime instance is created, and events at or after it—including events arriving while startup is delayed—must remain eligible. Moving it into start() would incorrectly fence that startup-gap traffic and violate the fixed-cutoff requirement.

File: src/mindroom/bot.py [hunk 2]
```diff
@@ -1134,6 +1148,59 @@ async def error_handler() -> None:
...
+            return True
+
+        origin_server_ts = origin_server_ts_from_event_source(event.source)
+        timestamp_is_finite = isinstance(origin_server_ts, int) or (
+            isinstance(origin_server_ts, float) and math.isfinite(origin_server_ts)
+        )
+        if timestamp_is_finite:
+            assert origin_server_ts is not None
+            if origin_server_ts >= self._startup_cutoff_ms:
+                return True
```
[the remark this answers is not in the document]
Comment (basnijholt) [reply src/mindroom/bot.py]: Not changing this. Matrix origin_server_ts and time.time() are both Unix-epoch milliseconds. Calibrating from the first cold-sync event cannot distinguish clock skew from genuine event age, and disabling the fence on apparent skew would fail open against the requested safety invariant. Missing or non-finite timestamps already fail closed.

File: src/mindroom/bot.py [hunk 3]
```diff
@@ -1134,6 +1148,59 @@ async def error_handler() -> None:
...
+            reason = "missing_origin_server_ts" if origin_server_ts is None else "invalid_origin_server_ts"
+
+        self.logger.info(
+            "matrix_initial_sync_dispatch_fenced",
+            callback=callback_name,
+            event_id=event.event_id,
+            room_id=room.room_id,
+            agent_name=self.agent_name,
+            reason=reason,
+        )
```
[the remark this answers is not in the document]
Comment (basnijholt) [reply src/mindroom/bot.py]: Fixed in 1cf395e38. Ordinary pre-cutoff replay logs at DEBUG; missing or invalid timestamps still log their required content-free reason at INFO. Added regressions for both levels.

File: tests/test_matrix_sync_tokens.py [hunk 4]
```diff
@@ -217,6 +260,383 @@ async def test_bot_start_restores_saved_sync_token(tmp_path: Path) -> None:
...
+        patch.object(bot, "_run_sync_response_side_effects", AsyncMock()),
+    ):
+        await bot._on_sync_response(response)
+
+    callback = cast("Callable[..., Awaitable[None]]", _registered_event_callback(client, nio.RoomMessageText))
+    handle_text_event = AsyncMock()
+    with patch.object(bot._turn_controller, "handle_text_event", handle_text_event):
+        await callback(room, event)
+        await wait_for_background_tasks(timeout=1.0, owner=bot._runtime_view)
+
```
[the remark this answers is not in the document]
Comment (basnijholt) [reply tests/test_matrix_sync_tokens.py]: Fixed in 25e4f909f. The regression now asserts the fence state at the owning sync-response seam after a missing continuation token. This removes the time, event, callback, and downstream dispatch mocks; only the two response collaborators are patched.

File: src/mindroom/bot.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (basnijholt) [reply src/mindroom/bot.py]: Not changing this. Matrix origin_server_ts, server_timestamp, and time.time() are Unix-epoch milliseconds. A first-event offset cannot distinguish homeserver clock skew from genuine historical age; a bounded allowance admits replay history; and disabling timestamp fencing would fail open. The approved invariant requires fail-closed handling, including missing or non-finite timestamps.

File: src/mindroom/bot.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (basnijholt) [reply src/mindroom/bot.py]: Not changing this. The approved requirement explicitly defines one immutable process-start cutoff captured when the bot/runtime instance is created. Events arriving after construction, including during delayed startup, are intentionally eligible. Moving the cutoff into start() would violate that required boundary and the fixed-cutoff retry invariant.

**Comment by basnijholt** [comment]
Closing for redesign after independent exact-head review.

The cold-history dispatch problem is high priority, but the central boundary here is invalid:

- local `time.time()` cannot safely order federated Matrix `origin_server_ts`; homeserver clock skew can admit old events or drop fresh ones;
- callback failure can clear the checkpoint before a durable pending turn exists;
- a later cold fence can then suppress that callback forever, including reactions and approvals that lack a generic pending-turn record.

Correct order:

1. Establish a durable exact callback/dispatch-obligation protocol.
2. During a sync window without trusted `since` or Sliding `pos`, suppress arbitrary history except exact durable obligations.
3. Resume ordinary dispatch only after a Matrix-issued continuation is established.
4. Run PR #1731's real restart profile against the combined candidate head.

Preserve useful regression concepts, but do not carry forward the wall-clock fence.
